### PR TITLE
Add lifecycle events emitting

### PR DESCRIPTION
**Pull request by hunger-programmer**

### CODE REVIEW

File: src/eventhub.ts [hunk 1]
```diff
@@ -19,6 +19,7 @@ THE SOFTWARE.
 */
 
 import WebSocket from 'isomorphic-ws';
+import events from "events";
```
Comment (bukowskiadam) [inline src/eventhub.ts]: Please keep consistent quotes. And import the new way - not the old compat way.
```suggestion
import EventEmitter from 'events';
```
https://github.com/browserify/events/blob/master/events.js#L60

File: src/eventhub.ts [hunk 2]
```diff
@@ -34,6 +35,13 @@ enum RPCMethods {
   DISCONNECT      = 'disconnect'
 }
 
+enum LifecycleEvents {
+  CONNECT         = "connect",
```
Comment (bukowskiadam) [inline src/eventhub.ts]:
```suggestion
  CONNECT         = 'connect',
```

File: src/eventhub.ts [hunk 3]
```diff
@@ -54,7 +62,11 @@ class ConnectionOptions {
   disablePingCheck: boolean = false;
 }
 
-export default class Eventhub {
+declare interface IEventhub {
```
Comment (bukowskiadam) [inline src/eventhub.ts:65]: Should we import this from `eventhub.d.ts`?

File: test/eventhub.test.ts [hunk 4]
```diff
@@ -129,3 +135,11 @@ test("Test that unsubscribeAll unsubscribes all subscribed topics", () => {
   expect(eventhub.isSubscribed("testTopic3")).toEqual(false);
   expect(eventhub.isSubscribed("testTopic4")).toEqual(false);
 });
+
+test("Test that reconnect event is emitted", () => {
```
Comment (bukowskiadam) [inline test/eventhub.test.ts]:
```suggestion
test('Test that reconnect event is emitted', () => {
```

File: src/eventhub.ts [hunk 5]
```diff
@@ -54,7 +62,11 @@ class ConnectionOptions {
...
+  on(event: LifecycleEvents, listener: Function): this;
+}
+
+export default class Eventhub extends events.EventEmitter implements IEventhub {
```
Comment (bukowskiadam) [inline src/eventhub.ts]:
```suggestion
export default class Eventhub extends EventEmitter implements IEventhub {
```

**Review by bukowskiadam — CHANGES_REQUESTED** [review]

File: test/eventhub.test.ts [hunk 6]
```diff
@@ -129,3 +135,11 @@ test("Test that unsubscribeAll unsubscribes all subscribed topics", () => {
...
+  testServer.close();
+
+  setTimeout(()=>{
+    expect(emitEventSpy).toHaveBeenCalledWith('reconnect');
+  }, 100)
```
Comment (filipgolonka) [inline test/eventhub.test.ts:148]: correct me if I'm wrong, but  this test will be always green, cause jest does not wait for async callbacks to be completed. You need to either wrap it with Promise or use `done` callback to say explicitly when the test is done

File: src/eventhub.ts (hunk 3 above)
Comment (bukowskiadam) [reply src/eventhub.ts:65]: sorry, my bad - that file is autogenerated in the dist directory 🙇

File: test/eventhub.test.ts (hunk 6 above)
Comment (hunger-programmer) [reply test/eventhub.test.ts:148]: 🙇 you are right, fixed!

**Review by bukowskiadam — APPROVED** [review]

**Review by olesku — APPROVED** [review]

**Pull request merged**